netebpfext/net_ebpf_ext.h

```diff
@@ -25,8 +25,29 @@
 #define NET_EBPF_EXTENSION_POOL_TAG 'Nfbe'
 #define NET_EBPF_EXTENSION_NPI_PROVIDER_VERSION 0
 
+#define NET_EBPF_EXT_MAX_CLIENTS_PER_HOOK_MULTI_ATTACH 16
```

Comment:
Why do we need a maximum number of clients?

Reply:
When the callout is invoked, we acquire filter context lock and create a copy of all the clients that are attached to the hook, as we want to invoke the programs outside the filter context hook. Keeping the max client count dynamic means we will need to allocate memory for every hook invocation (to create a copy of the clients), and that will be expensive. Keeping the max limit fixed allows allocating this memory on stack.

Reply:
Please add sufficient comments into the file, since I expect this will need to be changed someday (the day someone has a scenario for 17).

Reply:
Added.

Reply:
Re "Keeping the max client count dynamic means we will need to allocate memory for every hook invocation"... I don't think that logically follows, since it asserts the non-existence of any possible design that allows increasing the max count without requiring an allocation on every hook invocation.   I don't believe such a non-existence assertion.  It may be extra work and not high priority at the moment, but I suspect the sentence is provably false as currently phrased.

Reply:
I meant to say that keeping the max limit a compile time constant allows us to allocate the required memory on stack. The max limit today is set to 16. The design does not prohibit the current max limit to be increased to a new constant value in future, as long it is not too big to not being able to allocate on stack.

Reply:
Having a max value at compile time means some scenarios (hopefully ones we don't care about now) will fail.  It does not follow that keeping the max count dynamic means every possible implementation requires allocating memory on every hook invocation.  Counter examples include:
a) using alloca to dynamically allocate stack memory (does that work in kernel mode)?
b) pre-allocating per-cpu space dynamically and having a way to realloc it if needed (say as part of an epoch change)
etc.

I am not saying we have to do that now. However I am saying we should have an explanation in comments, while not having incorrect comments :)

Reply:
Agree that we can have other implementations that may not require dynamic memory allocation. I will update the text.

> pre-allocating (not at hook invocation time) per-cpu space and having a way to realloc it if needed (say as part of an epoch change)

To give more context about the above option, if the WFP callout is always invoked at DISPATCH, we dont need this logic of copying the clients in a local list. But the callout can be invoked either at DISPATCH or PASSIVE depending on the protocol.  Earlier I implemented this using a RW lock, but that now always raised the irql to DISPATCH before program invocation and was a kind of regression.

> using alloca to dynamically allocate stack memory (does that work in kernel mode)?

Another option can be to use epoch based memory management in netebpfext, and that should remove the need to create a copy,

Reply:
Updated